reference/src/layout/enums.md

```diff
@@ -313,13 +317,22 @@ values, which are called **niches**. For example, a value of type `&T`
 may never be `NULL`, and hence defines a niche consisting of the
 bitstring `0`.  Similarly, the standard library types [`NonZeroU8`]
 and friends may never be zero, and hence also define the value of `0`
-as a niche. (Types that define niche values will say so as part of the
-description of their validity invariant, which -- as of this writing
--- are the next topic up for discussion in the unsafe code guidelines
-process.)
+as a niche. 
 
 [`NonZeroU8`]: https://doc.rust-lang.org/std/num/struct.NonZeroU8.html
 
+The niche values of a type are parts of its validity invariant which, as of this
```

Comment:
The niche is not "part of the invariant", I don't even know what that means.

The niche values are *some of* values that variables of this type can never have. For example, a `&mut u32` can also never have value 1337, and yet it is unlikely that that will be in the niche. Currently it certainly is not.

If you want to draw a connection to validity, you can say that the niche values must be disjoint from the values allowed by the validity invariant.

Reply:
Oh and if you are anyway defining here what "niche" means, wouldn't it make sense to put that into the glossary?

Reply:
> If you want to draw a connection to validity, you can say that the niche values must be disjoint from the values allowed by the validity invariant

Done.

> a &mut u32 can also never have value 1337, and yet it is unlikely that that will be in the niche. 

Why is this unlikely? Not all invalid values can be niches (e.g. the "all-bits-uninitialized" bit-pattern), but for references, unaligned addresses are invalid and could be niches. I don't see a reason why they shouldn't be.

Reply:
> what "niche" means, wouldn't it make sense to put that into the glossary?

I'm not sure how to briefly define niche. Maybe something like "An invalid bit-pattern that can be used by layout optimizations.

Reply:
> Why is this unlikely? Not all invalid values can be niches (e.g. the "all-bits-uninitialized" bit-pattern), but for references, unaligned addresses are invalid and could be niches. I don't see a reason why they shouldn't be.

It doesn't fit the current architecture very well.
I guess long-term there's no problem with it, and certainly not in terms of spec, but short-term it seems unlikely.

> I'm not sure how to briefly define niche.

You are doing that already, right here. ;)

> Maybe something like "An invalid bit-pattern that can be used by layout optimizations".

Or "*will* be used"? That's what differentiates it from bit patterns that are ruled out by the validity invariant but *not* in the niche.

Reply:
I've added a definition of niche to the glossary. I'm resolving this, we can discuss that definition there.